### DISCUSSION TITLE
Add `assert` statement?

### DISCUSSION THREAD

**Discussion by rsmmr** [zeek/zeek#1792] [Ideas]
The discussion in #1475 on integrating ZTest led to some of us wondering if we really understand already how unit testing for Zeek scripts should look like. We can certainly continue that discussion, but there's a piece in there where I'm wondering if that be a good initial step to tackle independently first: adding an `assert` statement to Zeek's scripting language. Here's a straw man proposal on how that could look like:

1. Add a statement `assert <expr> [: <failure message>]`. `<expr>` is a boolean expression that is expected to be true. `<failure message>` is an optional string to report as an error message in case the boolean ends up being false.

2. By default, a failing `assert` produces an error through the reporter, including the failure message if given. It then aborts the current event handler by raising an `InterpreterException` (i.e., Zeek will continue its processing).

3. To change that default behavior, we provide a script-level hook `assert_failure` that runs whenever an assert fails, receiving (1) a string rendering of the failing expression, (2) the failure message, (3) location information where it happened. The hook would return a boolean to indicate whether it wants the script processing to continue normally despite the failing assert or abort through the `InterpreterExcption`. If it wanted, the hook could also just abort Zeek altogether at that point.
 
I believe that this would be helpful on its own for both (1) writing tests (which instead of printing/diffing could now write `assert ...`), and (2) normal scripts to enforce constraints they rely on. And if we later added a larger testing framework, it could continue to rely on `assert` and implement the hook for any custom processing (e.g., logging & counting failures).

What do people think?

**Comment by vpax** [comment]
In general, I like this notion.  It does though make me wonder about a more general exception mechanism, out of which one could build this.  (I have mixed feelings on whether that would be a good idea.)

**Comment by duffy-ocraven** [comment]
Although using the same mechanism, and fine to produce the same output, I would advocate for supplying BOTH `assert` and `requires`. Authors of source code tend to think "I can use assert only if I provide for every contingency". We also need the other end of the spectrum "handling an edge case is not worth my time now, but I should not silently act like this code handles it when it doesn't".

Use `requires` for things like not yet handling 0 passed to a function that takes the reciprocal. 
```
    returnTheReciprocal(inpVal)
        requires 0 <> inpVal
        return 1/inpVal
```
Whenever the code gets around to handling that in the intended graceful manner, the `requires` statement that had been excluding it, is removed during the same commit.
